File: frontend/ui/apps/pai_mobile/android/app/src/main/kotlin/com/example/pai_mobile/agent/plugins/NativeModuleRegistry.kt
package com.pai.agent.plugins

import java.util.concurrent.ConcurrentHashMap

class NativeModuleRegistry {
    private val modules = ConcurrentHashMap<String, PluginModule>()

    fun register(module: PluginModule) { modules[module.name] = module }
    fun unregister(name: String) { modules.remove(name) }
    fun get(name: String): PluginModule? = modules[name]
    fun list(): List<String> = modules.keys.toList()
    fun contains(name: String): Boolean = modules.containsKey(name)
}

File: frontend/pai_mobile/linux/runner/agent/plugins/native_module_registry.cc
// linux/runner/agent/plugins/native_module_registry.cc
#include "native_module_registry.h"

namespace pai::agent {

void NativeModuleRegistry::Register(std::shared_ptr<PluginModule> m) {
  if (!m) return;
  modules_[m->Name()] = std::move(m);
}

void NativeModuleRegistry::Unregister(const std::string& name) {
  modules_.erase(name);
}

std::shared_ptr<PluginModule> NativeModuleRegistry::Get(const std::string& name) const {
  auto it = modules_.find(name);
  return it == modules_.end() ? nullptr : it->second;
}

std::vector<std::string> NativeModuleRegistry::List() const {
  std::vector<std::string> out;
  out.reserve(modules_.size());
  for (const auto& [k, _] : modules_) out.push_back(k);
  return out;
}

bool NativeModuleRegistry::Contains(const std::string& name) const {
  return modules_.find(name) != modules_.end();
}

}  // namespace pai::agent